grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy

```diff
@@ -256,4 +253,67 @@ class PrepaymentInvoiceService {
     }
+
+    /**
+     * Bulk editing invoice items quantity on final invoice for prepaid POs.
+     * We are allowing to edit only the quantity here. After editing invoice items quantity we have to
+     * change quantity on the corresponding inverse item.
+     *
+     * */
+    void updateItemsQuantity(String invoiceId, List items) {
+        Invoice invoice = Invoice.get(invoiceId)
+        if (!invoice) {
+            throw new IllegalArgumentException("Cannot find invoice with id: ${invoiceId}")
+        }
+        if (invoice.isPrepaymentInvoice || !invoice.hasPrepaymentInvoice) {
+            throw new IllegalArgumentException("Cannot update quantities on prepayment or non prepaid invoices")
+        }
+
+        items.each { Map item ->
+            updateInvoiceItemQuantity(item.id, item.quantity)
+        }
+    }
+
+    /**
+     * Updates invoice item's quantity and amount and finds related inverse item and updates quantity and amount
+     * */
+    void updateInvoiceItemQuantity(String itemId, Integer quantity) {
+        InvoiceItem invoiceItem = InvoiceItem.get(itemId)
+        if (!invoiceItem) {
+            throw new IllegalArgumentException("Cannot find invoice with id: ${itemId}")
+        }
+        if (quantity <= 0) {
+            throw new IllegalArgumentException("Quantity to change needs to be higher than 0")
+        }
+        if (!invoiceItem?.shipmentItem) {
+            throw new IllegalArgumentException("Invoice item is missing shipment item. Cannot edit quantity.")
+        }
+        Integer quantityAvailableToInvoice = invoiceItem.shipmentItem.quantityToInvoice
+        if (quantity > quantityAvailableToInvoice + invoiceItem.quantity) {
+            throw new IllegalArgumentException("Cannot update quantity to higher value than available to invoice")
+        }
+        // update invoice item's quantity (and adjust amount)
+        invoiceItem.quantity = quantity
+        invoiceItem.amount = quantity * invoiceItem.unitPrice
+        // update inverse item's quantity if there is inverse item (and adjust amount)
+        InvoiceItem inverseItem = findInverseItem(invoiceItem)
+        if (inverseItem) {
+            OrderItem orderItem = invoiceItem?.shipmentItem?.orderItems?.find { it }
+            InvoiceItem prepaymentItem = orderItem.invoiceItems.find { it.isPrepaymentInvoice }
+            // find quantity that is still available to inverse and add to the current quantity from this inverse item
+            // this needs to be checked because we can also increase invoiced quantity here
+            Integer quantityAvailableToInverse = inverseItem.quantity + getQuantityAvailableToInverse(orderItem, prepaymentItem)
+            Integer quantityToInverse = quantity > quantityAvailableToInverse ? quantityAvailableToInverse : quantity
+            inverseItem.quantity = quantityToInverse
+            inverseItem.amount = quantityToInverse * prepaymentItem.unitPrice * orderItem.order.prepaymentPercent * (-1)
```

Comment:
@EWaterman I still remember you wanted to move it to separate method but I was not sure how to actually do it. Is something like that what you'd expect here:
```groovy
BigDecimal calculateAmount(Integer quantity, BigDecimal unitPrice, BigDecimal percentage = 1, Integer inverse = 1) {
    return quantity * unitPrice * percentage * inverse
}
```
and then just passing what we have available or need to use in calculation, or do you see here something a bit more complex in this method?

Reply:
I think the thing I'm mainly worried about is that we calculate amount all over the place and it'd be nice to have a calculateAmountForPrepaymentItem(), calculateAmountForRegularItem() and calculateAmountForInverseItem() methods just to have it unified.

I haven't thought it through fully so this might not be the best idea, but they could maybe just take in the invoice item itself and do all the lookups they need so that the caller doesn't need to know what to do.

The logic here is just a bit tough to navigate with all the transients and chained calls so it'd be great if there was a simple method of calculating it that brought it all together

Reply:
Ok, I think I am going to leave it as is for now, but I'll think about this as a standalone improvement in separate ticket/pr